libraries/resource_providers/api200/volume_template_provider.rb

```diff
@@ -13,28 +13,29 @@ module OneviewCookbook
   module API200
     # VolumeTemplate API200 provider
     class VolumeTemplateProvider < ResourceProvider
+      include OneviewCookbook::Validations::Presence
+
       # Loads the VolumeTemplate with all the external resources (if needed)
       def load_resource_with_associated_resources
-        raise "Unspecified property: 'storage_system'. Please set it before attempting this action." unless @new_resource.storage_system
-        raise "Unspecified property: 'storage_pool'. Please set it before attempting this action." unless @new_resource.storage_pool
+        validate_presence_of(:storage_system, :storage_pool)
+        storage_system = load_storage_system
+        storage_pool = load_resource(:StoragePool, name: @new_resource.storage_pool, storageSystemUri: storage_system['uri'])
+
+        @item.set_storage_system(storage_system)
+        @item['provisioning']['storagePoolUri'] = storage_pool['uri']
         @item['provisioning']['capacity'] = @item['provisioning']['capacity'].to_s if @item['provisioning'] && @item['provisioning']['capacity']
-        load_storage_system
-        sp = resource_named(:StoragePool).find_by(@item.client, name: @new_resource.storage_pool, storageSystemUri: @item['storageSystemUri']).first
-        raise "Storage Pool '#{@new_resource.storage_pool}' not found for Storage System '#{@new_resource.storage_system}'" unless sp
-        @item['provisioning']['storagePoolUri'] = sp['uri']
-        @item.set_snapshot_pool(resource_named(:StoragePool).new(@item.client, name: @new_resource.snapshot_pool)) if @new_resource.snapshot_pool
+        @item.set_snapshot_pool(load_resource(:StoragePool, name: @new_resource.snapshot_pool, storageSystemUri: storage_system['uri'])) if @new_resource.snapshot_pool
```

Comment:
In this case, since the StoragePool has been loaded before, using `load_resource` again seems very expensive for no gain.
Likely it should be:
```ruby
@item.set_snapshot_pool(storage_pool) if @new_resource.snapshot_pool
```

Reply:
But the problem is that the `snapshot_pool` can have a different value.
I can test if `storage_pool` and `snapshot_pool` have the same value, but I think that it is more simple verify just if  `snapshot_pool` is present and find it instead of verify if `snapshot_pool` is present and verify if the value is the same of `storage_pool`, and apply different actions according with the result of verification. What do you think?

Reply:
I just want to understand, do we **need** to call oneview **here** for some reason?
The code being replaced:
```ruby
@item.set_snapshot_pool(resource_named(:StoragePool).new(@item.client, name: @new_resource.snapshot_pool)) if @new_resource.snapshot_pool
```
Just instantiates the `StoragePool` resource and uses it as input for `set_snapshot_pool`

The new line:
```ruby
@item.set_snapshot_pool(load_resource(:StoragePool, name: @new_resource.snapshot_pool, storageSystemUri: storage_system['uri'])) if @new_resource.snapshot_pool
```
is making a call to oneview. Is that call necessary by any means?

Reply:
Also, in [line 22](https://github.com/HewlettPackard/oneview-chef/pull/328/files/fc89477c7c23841a0025f9256bc8e93b908b0855#diff-3117eecb1f99d10e576f18b8e38aaaa9R22):
```ruby
storage_pool = load_resource(:StoragePool, name: @new_resource.storage_pool, storageSystemUri: storage_system['uri'])
```
vs [line 37](https://github.com/HewlettPackard/oneview-chef/pull/328/files/fc89477c7c23841a0025f9256bc8e93b908b0855#diff-3117eecb1f99d10e576f18b8e38aaaa9R27):
```ruby
@item.set_snapshot_pool(load_resource(:StoragePool, name: @new_resource.snapshot_pool, storageSystemUri: storage_system['uri'])) if @new_resource.snapshot_pool
``` 
Since the `load_resource` call is exactly the same, and I don't see the variable `storage_pool` being modified, why would it not be OK to use it here?

Reply:
Nvm, I'm fine with this.